### DISCUSSION TITLE
How to create distinct types from a single source type?

### DISCUSSION THREAD

**Discussion by Dimension4** [dotnet/csharplang#7753] [Q&A]
Hey there,

as far as I know, C# doesn't allow you to create distinct types (see [Nim](https://nim-by-example.github.io/types/distinct/), for example). Other than keeping multiple identical implementations with different class / struct names, what options do I have? Does anybody know about an existing source generator that does something in that direction?

Assume I have a `Vector3` type. I usually work in two different coordinate systems. Now using the same type for all liner algebra can be a bit tricky since I always have to make sure that all terms in an equation are operating in the same coordinate system (unless I want to convert between the two). It would be nice to have two distinct types that only convert explicitly to each other. Say I have the following function that correlates two vectors in my two coordinate systems:

```cs
void Correlate(Vector3 pointInSystemA, Vector3 pointInSystemB);
```
 In this case, the compiler can't help me if I mix up the argument order or pass in two vectors from the same coordinate system. With distinct types, I could do the following:

```cs
using Vector3A = distinct Vector3;
using Vector3B = distinct Vector3;

Vector3A x = new(1, 2, 3);
Vector3B y = new(4, 5, 6);
Vector3 z = new(0, 0, 0);

void Correlate(Vector3A pointInSystemA, Vector3B pointInSystemB)
{
    // ...
}

Correlate(x, y); // compiles
Correlate(y, x); // error
Correlate(x, (Vector3B) x); // compiles (or not, not really important)
Correlate(x, z); // error

//
// ideally, the "source" type still accepts other vectors as arguments (maybe via implicit conversions)
//

void SomeMethodThatOperatesOnOrdinaryVectors(Vector3 x)
{
    // ...
}

SomeMethodThatOperatesOnOrdinaryVectors(x); // compiles
SomeMethodThatOperatesOnOrdinaryVectors(y); // compiles
SomeMethodThatOperatesOnOrdinaryVectors(z); // compiles
```

**Comment by huoyaoyuan** [comment]
This is a design goal of explicit extension. See #5497 https://github.com/dotnet/csharplang/discussions/7094#discussioncomment-5536560
